### PR TITLE
GH-747 Adds Hover Indicators to layout/Cell.js

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH-747

@GordonSmith @dtsnell4 @mlzummo I've added **indicateTheseIds** publish param to layout/Cell.js

It holds an array of DOM IDs and displays a div over each corresponding html element when the use hovers over the Cell.

https://rawgit.com/jbrundage/Visualization/GH-747-HoverIndicators/demos/dermatology.html?src/layout/Grid.hoverIndicator

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
That is not really what I am expecting (or might be a bug) - all should look normal, but when I hover over one cell, it should subtly change color(say very very pale yellow) and then the "indicateTheseIDs" list should be highlighted in a slightly less subtle shade of yellow (one of the standard "highlighter" colors).

File: src/layout/Cell.js [hunk 1]
```diff
@@ -19,16 +19,29 @@
...
+    Cell.prototype.publish("indicatorColor", "#FF00FF", "html-color", "Background color of update-indicator",null,{tags:["Basic"]});
+    Cell.prototype.publish("indicatorOpacity", 0.2, "number", "Opacity of update-indicator",null,{tags:["Basic"]});
 
     Cell.prototype.enter = function (domNode, element) {
         Surface.prototype.enter.apply(this, arguments);
         element.classed("layout_Surface", true);
+        
+        //Example values...
+        this.indicateTheseIds(["_w43","_w59"]);
+
```
Comment (GordonSmith) [inline src/layout/Cell.js]: Can you move this into its own Test in test/layoutFactory - I suspect the issue I think I am seeing is becuase this is being set for every cell...

Comment (jbrundage) [inline src/layout/Cell.js]: @GordonSmith Done

**Comment by GordonSmith** [comment]
Ok that is working better, I think the highlighting should be done at the cell/surface border level rather than the content - thoughts?  Here is what I am seeing - I would expect the titlebar and border to be used to indicate and leave the content alone:
image

**Comment by jbrundage** [comment]
@GordonSmith The coloring is just a fixed div placed overtop of whichever ID it's given. If it should be applied at the Cell/Surface level the method that specifies which widgets need to be highlighted could just return the IDs of the Cell/Surface.

IMO if we modify the Surface styles directly it's a bit messier of an implementation (and puts limitations on how this feature can be used going forward).

**Comment by dtsnell4** [comment]
What if you just removed the background color of the div, expanded it to the size of the Surface container and changed it to a box-shadow to give it an "outer glow" effect?

**Comment by jbrundage** [comment]
@dtsnell4 That would work well. I'll update this PR with that change...

File: src/layout/Cell.js [hunk 2]
```diff
@@ -139,6 +148,38 @@
...
+                div.className = "update-indicator";
+                div.style.position = "fixed";
+                div.style.top = rect.top + "px";
+                div.style.left = rect.left + "px";
+                div.style.width = rect.width + "px";
+                div.style.height = rect.height + "px";
+                div.style.opacity = opacity;
+                div.style.padding = "0px";
+                div.style.zIndex = 1000;
+                div.style.backgroundColor = bgColor;
```
Comment (GordonSmith) [inline src/layout/Cell.js]: I would rather you used ```d3.append().style({xxx:yyy});``` paradigm:
* d3 fixes a few cross browser inconsistencies 
* We could then use transitions if we wanted to
* general code consistency

File: src/layout/Cell.js [hunk 3]
```diff
@@ -139,6 +148,38 @@
...
+            }
+        };
+    };
+    Cell.prototype.onMouseLeave = function (){
+        var element = this._element;
```
Comment (GordonSmith) [inline src/layout/Cell.js]: ```this.element()``` is preferable.

**Comment by GordonSmith** [comment]
@jbrundage bump.

**Comment by jbrundage** [comment]
@GordonSmith @dtsnell4 I've made the suggested changes. Please review

https://rawgit.com/jbrundage/Visualization/GH-747-HoverIndicators/demos/dermatology.html?src/layout/Grid.hoverIndicator

**Comment by dtsnell4** [comment]
That looks good to me.

**Comment by mlzummo** [comment]
@jbrundage it doesnt always turn off the indication when you mouse out

**Comment by GordonSmith** [comment]
@jbrundage Code looks ok - I tried to tweak the colours in the dermatology page, but the new values didn't seem to change?

**Comment by jbrundage** [comment]
@GordonSmith I looked into it just now and this is a separate bug. The PropertyEditor is reversing the order of the "Cell" widgets. You would have to change the indicatorColor params in the bottom two "Cell" sections to see the color change.

**Comment by GordonSmith** [comment]
@jbrundage Can you open a new issue for that?

**Pull request merged**